File: code/SoundPlayground/SoundPlayground/Parser/AbstractSyntaxTree/ContextModifiers/OctaveModifierNode.cs
using System.Collections.Generic;
using SoundPlayground.Core;

namespace SoundPlayground.Parser.AbstractSyntaxTree.ContextModifiers {
    public class OctaveModifierNode : ContextModifierNode {
        public int Octave { get; set; }

        public OctaveModifierNode ( int octave ) {
            Octave = octave;
        }

        public override void Modify ( Context context ) {
            context.Octave = Octave;
        }
    }
}

File: code/SoundPlaygroundPy/parser/abstract_syntax_tree/context_modifiers/octave_modifier_node.py
from .context_modifier_node import ContextModifierNode
from core.events import ContextChangeEvent
from core import Context, Voice

class OctaveModifierNode( ContextModifierNode ):
    def __init__ ( self, octave, position : (int, int) = None ):
        super().__init__( position )

        self.octave = octave
        
    def apply ( self, voice : Voice ):
        voice.octave = self.octave

    def modify ( self, context : Context ):
        context.voice = context.voice.clone( octave = self.octave )

        yield ContextChangeEvent( context.cursor, "octave", context.voice.octave )
